core/src/scheduled_jobs/pending_validations.rs

```diff
@@ -1,26 +1,47 @@
 use crate::{
-    context::Context, network::entry_with_header::EntryWithHeader,
-    workflows::hold_link::hold_link_workflow,
+    context::Context,
+    network::entry_with_header::EntryWithHeader,
+    nucleus::actions::remove_pending_validation::remove_pending_validation,
+    workflows::{hold_entry::hold_entry_workflow, hold_link::hold_link_workflow},
+};
+use holochain_core_types::{
+    cas::content::{Address, AddressableContent},
+    entry::entry_type::EntryType,
+    error::error::HolochainError,
 };
-use holochain_core_types::{cas::content::Address, entry::entry_type::EntryType};
 use std::{sync::Arc, thread};
 
-pub type PendingValidation = Box<(EntryWithHeader, Vec<Address>)>;
+pub type PendingValidation = Arc<(EntryWithHeader, Vec<Address>)>;
 
 fn retry_validation(pending: PendingValidation, context: Arc<Context>) {
-    thread::spawn(move || context.block_on(hold_link_workflow(&pending.0, &context)));
+    thread::spawn(move || {
+        let result = match pending.0.entry.entry_type() {
+            EntryType::LinkAdd | EntryType::LinkRemove => {
+                context.block_on(hold_link_workflow(&pending.0, &context))
+            }
+            _ => context.block_on(hold_entry_workflow(&pending.0, context.clone())),
+        };
+
+        if Err(HolochainError::ValidationPending) != result {
+            remove_pending_validation(pending.0.entry.address(), &context);
+        }
+    });
 }
 
 pub fn run_pending_validations(context: Arc<Context>) {
-    context
+    let pending_validations = context
         .state()
         .unwrap()
         .nucleus()
         .pending_validations
-        .iter()
-        .for_each(|(_, boxed)| match boxed.0.entry.entry_type() {
-            EntryType::LinkAdd => retry_validation(boxed.clone(), context.clone()),
-            EntryType::LinkRemove => retry_validation(boxed.clone(), context.clone()),
-            _ => panic!("Pending validations are (currently) only implemented for links"),
-        });
+        .clone();
```

Comment:
We get this clone of a `HashMap<Arc<...>>` so we can drop the state lock immediately.